nixos/modules/services/networking/spruned.nix

```diff
@@ -0,0 +1,60 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.spruned;
+
+  cliArgs = "--datadir ${cfg.dataDir} --network bitcoin.${cfg.network} ${cfg.extraArguments}";
+in
+{
+  options = {
+    services.spruned = {
+      enable = mkOption {
+        type = types.bool;
+        default = false;
+        description = ''
+          Whether or not to enable the spruned lightweight Bitcoin pseudonode
+          daemon service.
+        '';
+      };
+      dataDir = mkOption {
+        type = types.string;
+        description = ''
+          Directory to store cached block data and spruned logs.
+        '';
+      };
+      network = mkOption {
+        type = types.enum [ "mainnet" "testnet" ];
+        default = "mainnet";
+        description = ''
+          Whether to use Bitcoin mainnet or testnet.
+        '';
+      };
+      extraArguments = mkOption {
+        type = types.string;
+        example = "--mempoolsize 20 --debug";
+        default = "";
+        description = ''
+          Additional arguments to be passed to spruned.
+        '';
+      };
+    };
+  };
+
+  config = mkIf cfg.enable {
+    systemd.services.spruned = {
+      description = "spruned service";
+      after = [ "local-fs.target" "network.target" ];
+      wantedBy = [ "multi-user.target" ];
+
+      # needs ping
+      path = [ pkgs.iputils ];
+
+      serviceConfig = {
+        Restart = "on-abort";
+        ExecStart = "${pkgs.spruned}/bin/spruned ${cliArgs}";
```

Comment:
Like this, the service runs as root. Try using the `DynamicUser` and `StateDirectory` systemd settings (`man systemd.exec`) to prevent that.

Reply:
so I tried running it as non-root by first using user services and then assigned users, but spruned calls out to ping and doesn't work without it. I couldn't figure out how to make a systemd service that uses ping and isn't root. I looked into the CapabilityBoundingSet CAP_NET_RAW thing but it didn't seem to work.

Reply:
Use `/run/wrappers/bin/ping` instead of the ping in `iputils`. And really try to use `DynamicUser`, I think it should work.

Reply:
ok I'll try to take another pass at this and patch spruned to use /run/wrappers/bin/ping and switch to DynamicUser. Why DynamicUser instead of a user service?

Reply:
DynamicUser lets you can run an un-elevated service without assigning it a system user id.

Reply:
I'm getting this when I try to run with DynamicUser:

```
Nov 17 12:42:17 monad systemd[1]: Started spruned service.
Nov 17 12:42:17 monad spruned[17804]: Failed to import the site module
Nov 17 12:42:17 monad spruned[17804]: Traceback (most recent call last):
Nov 17 12:42:17 monad spruned[17804]:   File "/nix/store/nrl0l79a48924xb0897ap572xf29ciir-python3-3.6.6/lib/python3.6/site.py", line 544, in <module>
Nov 17 12:42:17 monad spruned[17804]:     main()
Nov 17 12:42:17 monad spruned[17804]:   File "/nix/store/nrl0l79a48924xb0897ap572xf29ciir-python3-3.6.6/lib/python3.6/site.py", line 530, in main
Nov 17 12:42:17 monad spruned[17804]:     known_paths = addusersitepackages(known_paths)
Nov 17 12:42:17 monad spruned[17804]:   File "/nix/store/nrl0l79a48924xb0897ap572xf29ciir-python3-3.6.6/lib/python3.6/site.py", line 282, in addusersitepackages
Nov 17 12:42:17 monad spruned[17804]:     user_site = getusersitepackages()
Nov 17 12:42:17 monad spruned[17804]:   File "/nix/store/nrl0l79a48924xb0897ap572xf29ciir-python3-3.6.6/lib/python3.6/site.py", line 258, in getusersitepackages
Nov 17 12:42:17 monad spruned[17804]:     user_base = getuserbase() # this will also set USER_BASE
Nov 17 12:42:17 monad spruned[17804]:   File "/nix/store/nrl0l79a48924xb0897ap572xf29ciir-python3-3.6.6/lib/python3.6/site.py", line 248, in getuserbase
Nov 17 12:42:17 monad spruned[17804]:     USER_BASE = get_config_var('userbase')
Nov 17 12:42:17 monad spruned[17804]:   File "/nix/store/nrl0l79a48924xb0897ap572xf29ciir-python3-3.6.6/lib/python3.6/sysconfig.py", line 601, in get_config_var
Nov 17 12:42:17 monad spruned[17804]:     return get_config_vars().get(name)
Nov 17 12:42:17 monad spruned[17804]:   File "/nix/store/nrl0l79a48924xb0897ap572xf29ciir-python3-3.6.6/lib/python3.6/sysconfig.py", line 558, in get_config_vars
Nov 17 12:42:17 monad spruned[17804]:     _CONFIG_VARS['userbase'] = _getuserbase()
Nov 17 12:42:17 monad spruned[17804]:   File "/nix/store/nrl0l79a48924xb0897ap572xf29ciir-python3-3.6.6/lib/python3.6/sysconfig.py", line 205, in _getuserbase
Nov 17 12:42:17 monad spruned[17804]:     return joinuser("~", ".local")
Nov 17 12:42:17 monad spruned[17804]:   File "/nix/store/nrl0l79a48924xb0897ap572xf29ciir-python3-3.6.6/lib/python3.6/sysconfig.py", line 184, in joinuser
Nov 17 12:42:17 monad spruned[17804]:     return os.path.expanduser(os.path.join(*args))
Nov 17 12:42:17 monad spruned[17804]:   File "/nix/store/nrl0l79a48924xb0897ap572xf29ciir-python3-3.6.6/lib/python3.6/posixpath.py", line 249, in expanduser
Nov 17 12:42:17 monad spruned[17804]:     userhome = pwd.getpwuid(os.getuid()).pw_dir
Nov 17 12:42:17 monad spruned[17804]: KeyError: 'getpwuid(): uid not found: 61833'
Nov 17 12:42:17 monad systemd[1]: spruned.service: Main process exited, code=exited, status=1/FAILURE
Nov 17 12:42:17 monad systemd[1]: spruned.service: Failed with result 'exit-code'.
```